File: ontrack-extension-jenkins/src/main/java/net/nemerosa/ontrack/extension/jenkins/JenkinsHealthIndicator.java
package net.nemerosa.ontrack.extension.jenkins;

import net.nemerosa.ontrack.extension.jenkins.client.JenkinsClientFactory;
import net.nemerosa.ontrack.model.security.SecurityService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.actuate.health.Health;
import org.springframework.boot.actuate.health.HealthAggregator;
import org.springframework.boot.actuate.health.HealthIndicator;
import org.springframework.stereotype.Component;

import java.util.Map;
import java.util.stream.Collectors;

@Component
public class JenkinsHealthIndicator implements HealthIndicator {

    private final JenkinsConfigurationService configurationService;
    private final SecurityService securityService;
    private final HealthAggregator healthAggregator;
    private final JenkinsClientFactory jenkinsClientFactory;

    @Autowired
    public JenkinsHealthIndicator(JenkinsConfigurationService configurationService, SecurityService securityService, HealthAggregator healthAggregator, JenkinsClientFactory jenkinsClientFactory) {
        this.configurationService = configurationService;
        this.securityService = securityService;
        this.healthAggregator = healthAggregator;
        this.jenkinsClientFactory = jenkinsClientFactory;
    }

    @Override
    public Health health() {
        return securityService.asAdmin(() -> {
            Map<String, Health> healths = configurationService.getConfigurations().stream()
                    .collect(
                            Collectors.toMap(
                                    JenkinsConfiguration::getName,
                                    this::getHealth
                            )
                    );
            return healthAggregator.aggregate(healths);
        });
    }

    private Health getHealth(JenkinsConfiguration config) {
        try {
            jenkinsClientFactory.getClient(config).getInfo();
            return Health.up().build();
        } catch (Exception ex) {
            return Health.down(ex).build();
        }
    }
}

File: ontrack-extension-jenkins/src/main/java/net/nemerosa/ontrack/extension/jenkins/JenkinsHealthIndicator.kt
package net.nemerosa.ontrack.extension.jenkins

import net.nemerosa.ontrack.extension.jenkins.client.JenkinsClientFactory
import net.nemerosa.ontrack.extension.support.ConfigurationConnectorStatusIndicator
import net.nemerosa.ontrack.model.security.SecurityService
import net.nemerosa.ontrack.model.support.ConfigurationService
import net.nemerosa.ontrack.model.support.ConnectorDescription
import net.nemerosa.ontrack.model.support.ConnectorStatus
import org.springframework.stereotype.Component

@Component
class JenkinsHealthIndicator(
        configurationService: ConfigurationService<JenkinsConfiguration>,
        securityService: SecurityService,
        private val jenkinsClientFactory: JenkinsClientFactory
) : ConfigurationConnectorStatusIndicator<JenkinsConfiguration>(configurationService, securityService) {

    override fun getConnectorStatus(config: JenkinsConfiguration): ConnectorStatus {
        return try {
            jenkinsClientFactory.getClient(config).info
            ConnectorStatus.ok(connectorDescription(config))
        } catch (ex: Exception) {
            ConnectorStatus.error(connectorDescription(config), ex)
        }

    }

    private fun connectorDescription(config: JenkinsConfiguration) = ConnectorDescription(
            type = "jenkins",
            name = config.name,
            connection = config.url
    )
}
